tonic/src/transport/service/router.rs

```diff
@@ -0,0 +1,128 @@
+use futures_util::{
+    future::Either,
+    try_future::{MapErr, TryFutureExt},
+};
+use std::{
+    fmt,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower_service::Service;
+
+#[derive(Debug)]
+pub(crate) struct Routes<A, B, Request> {
+    routes: Or<A, B, Request>,
+}
+
+impl<A, B, Request> Routes<A, B, Request> {
+    pub(crate) fn new(
+        predicate: impl Fn(&Request) -> bool + Send + Sync + 'static,
+        a: A,
+        b: B,
+    ) -> Self {
+        let routes = Or::new(predicate, a, b);
+        Self { routes }
+    }
+}
+
+impl<A, B, Request> Routes<A, B, Request> {
+    pub(crate) fn push<C>(
+        self,
+        predicate: impl Fn(&Request) -> bool + Send + Sync + 'static,
```

Comment:
It might be worth considering a `Predicate` trait so that users can add more complex filtering? Probably not in this branch though.

Reply:
Yeah, once we get a stable release of tokio, im gonna update a bunch of the tower services and will impl this there.